tests/src/Functional/TypedDataFormWidget/TextareaWidgetTest.php

```diff
@@ -0,0 +1,89 @@
+<?php
+
+namespace Drupal\Tests\typed_data\Functional\TypedDataFormWidget;
+
+use Drupal\Core\Plugin\Context\ContextDefinition;
+use Drupal\Core\TypedData\DataDefinition;
+use Drupal\Core\TypedData\ListDataDefinition;
+use Drupal\Core\TypedData\MapDataDefinition;
+use Drupal\Core\TypedData\TypedDataTrait;
+use Drupal\Tests\BrowserTestBase;
+use Drupal\Tests\typed_data\Traits\BrowserTestHelpersTrait;
+use Drupal\typed_data\Util\StateTrait;
+use Drupal\typed_data\Widget\FormWidgetManagerTrait;
+
+/**
+ * Class TextInputWidgetTest.
+ *
+ * @group typed_data
+ *
+ * @coversDefaultClass \Drupal\typed_data\Plugin\TypedDataFormWidget\TextareaWidget
+ */
+class TextareaWidgetTest extends BrowserTestBase {
+
+  use BrowserTestHelpersTrait;
+  use FormWidgetManagerTrait;
+  use StateTrait;
+  use TypedDataTrait;
+
+  /**
+   * The tested form widget.
+   *
+   * @var \Drupal\typed_data\Widget\FormWidgetInterface
+   */
+  protected $widget;
+
+  /**
+   * Modules to enable.
+   *
+   * @var array
+   */
+  public static $modules = [
+    'typed_data',
+    'typed_data_widget_test',
+  ];
+
+  /**
+   * {@inheritdoc}
+   */
+  public function setUp() {
+    parent::setUp();
+    $this->widget = $this->getFormWidgetManager()->createInstance('textarea');
+  }
+
+  /**
+   * @covers ::isApplicable
+   */
+  public function testIsApplicable() {
+    $this->assertTrue($this->widget->isApplicable(DataDefinition::create('string')));
+    $this->assertFalse($this->widget->isApplicable(ListDataDefinition::create('string')));
+    $this->assertFalse($this->widget->isApplicable(MapDataDefinition::create()));
```

Comment:
Thanks!
Fixed, but I am not sure should we leave the following code?
`$this->assertFalse($this->widget->isApplicable(ListDataDefinition::create('string')));
    $this->assertFalse($this->widget->isApplicable(MapDataDefinition::create()));`

Reply:
those two are fine, but the others should not be removed. There should be the whole list of data types in the test, so we can make sure the widget is available for the right types!

>
With
>I don't think this makes sense - a textarea for date strings does not fit very well. I think should be only be applicable to the 'string' data type for now.
I meant that we need to fix isApplicable() to return only true when it's a "string" data type.

Reply:
Sorry, I didn't get it.
Should we have the following testIsApplicable() ?

public function testIsApplicable() {
    $this->assertFalse($this->widget->isApplicable(DataDefinition::create('any')));
    $this->assertFalse($this->widget->isApplicable(DataDefinition::create('binary')));
    $this->assertFalse($this->widget->isApplicable(DataDefinition::create('boolean')));
    $this->assertFalse($this->widget->isApplicable(DataDefinition::create('datetime_iso8601')));;
    $this->assertFalse($this->widget->isApplicable(DataDefinition::create('duration_iso8601')));
    $this->assertFalse($this->widget->isApplicable(DataDefinition::create('email')));
    $this->assertFalse($this->widget->isApplicable(DataDefinition::create('float')));
    $this->assertFalse($this->widget->isApplicable(DataDefinition::create('integer')));
    $this->assertTrue($this->widget->isApplicable(DataDefinition::create('string')));
    $this->assertFalse($this->widget->isApplicable(DataDefinition::create('timespan')));
    $this->assertFalse($this->widget->isApplicable(DataDefinition::create('timestamp')));
    $this->assertFalse($this->widget->isApplicable(DataDefinition::create('uri')));
    $this->assertFalse($this->widget->isApplicable(ListDataDefinition::create('string')));
    $this->assertFalse($this->widget->isApplicable(MapDataDefinition::create()));
  }

Reply:
yes, exactly!